### PR TITLE
Add base optimizer

### PR DESCRIPTION
**Pull request by nnsgmsone**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #2372

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

**Review by ouyuanning — APPROVED** [review]

**Comment by nnsgmsone** [comment]
By the way, instead of calling optimize, the front-end now calls buildplan directly. This may need to be modified.

**Review by aunjgr — APPROVED** [review]

File: pkg/sql/plan2/optimize.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+func (opt *BaseOptimizer) optimize(qry *Query) (*Query, error) {
+	if len(qry.Steps) != 1 {
+		return nil, errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("query '%s' not support now", qry))
+	}
+	n := qry.Nodes[qry.Steps[0]]
+	opt.exploreNode(n)
+	for i := range n.Children {
+		opt.exploreNode(qry.Nodes[n.Children[i]])
+	}
```
[the remark this answers is not in the document]
Comment (nnsgmsone) [reply pkg/sql/plan2/optimize.go]: yes, fixed

File: pkg/sql/plan2/optimize.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	qry, ok := pn.Plan.(*plan.Plan_Query)
+	if !ok {
+		panic(errors.New(errno.SyntaxErrororAccessRuleViolation, pn.String()))
+	}
+	return opt.optimize(qry.Query)
+}
+
+func (opt *BaseOptimizer) optimize(qry *Query) (*Query, error) {
+	if len(qry.Steps) != 1 {
```
Comment (fengttt) [inline pkg/sql/plan2/optimize.go:71]: Loop over steps and optimize all steps.

Comment (nnsgmsone) [reply pkg/sql/plan2/optimize.go:71]: ok, mainly now plan2 doesn't generate steps greater than 1, so I want to panic to mask the undefined input.

File: pkg/sql/plan2/optimize.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+func (opt *BaseOptimizer) exploreNode(n *Node) {
+	rules, ok := opt.rules[n.NodeType]
+	if !ok {
+		return
+	}
+	for _, rule := range rules {
+		if rule.Match(n) {
+			rule.Apply(n, opt.qry)
+		}
+	}
```
Comment (fengttt) [inline pkg/sql/plan2/optimize.go:88]: Most likely you want to do post-order traversal.   That is, apply to children first, then self.

File: pkg/sql/plan2/optimize.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+		ctx:   ctx,
+		rules: defaultRules,
+	}
+}
+
+func (opt *BaseOptimizer) CurrentContext() CompilerContext {
+	return opt.ctx
+}
+
+func (opt *BaseOptimizer) Optimize(stmt tree.Statement) (*Query, error) {
```
Comment (fengttt) [inline pkg/sql/plan2/optimize.go:58]: Usually, optimizer do this, for each rule, walk the tree, then apply.

File: pkg/sql/plan2/optimize.go (hunk 3 above)
Comment (nnsgmsone) [reply pkg/sql/plan2/optimize.go:88]: ok

File: pkg/sql/plan2/optimize.go (hunk 4 above)
Comment (fengttt) [reply pkg/sql/plan2/optimize.go:58]: Therefore, usually it does not map node type to rule.   Instead, each rule, can handle certain types of nodes.

Comment (nnsgmsone) [reply pkg/sql/plan2/optimize.go:58]: ok

**Review by yingfeng — APPROVED** [review]

**Pull request merged**